src/abstractdatatable/join.jl

```diff
@@ -3,217 +3,170 @@
-    index = sort(unique([levels(v1); levels(v2)]))
-    sz = length(index)
-
-    R = sz <= typemax(UInt8)  ? UInt8 :
-        sz <= typemax(UInt16) ? UInt16 :
-        sz <= typemax(UInt32) ? UInt32 :
-                                UInt64
-
-    # To ensure type stability during actual work
-    sharepools(v1, v2, index, R)
-end
-
-sharepools{S,N}(v1::Union{CategoricalArray{S,N}, NullableCategoricalArray{S,N}},
-                v2::AbstractArray{S,N}) =
-    sharepools(v1, oftype(v1, v2))
-
-sharepools{S,N}(v1::AbstractArray{S,N},
-                v2::Union{CategoricalArray{S,N}, NullableCategoricalArray{S,N}}) =
-    sharepools(oftype(v2, v1), v2)
-
-# TODO: write an optimized version for (Nullable)CategoricalArray
-function sharepools{S, T}(v1::AbstractArray{S},
-                          v2::AbstractArray{T})
-    ## Return two categorical arrays that share the same pool.
-
-    ## TODO: allow specification of R
-    R = CategoricalArrays.DefaultRefType
-    refs1 = Array{R}(size(v1))
-    refs2 = Array{R}(size(v2))
-    K = promote_type(S, T)
-    poolref = Dict{K, R}()
-    maxref = 0
-
-    # loop through once to fill the poolref dict
-    for i = 1:length(v1)
-        if !_isnull(v1[i])
-            poolref[K(v1[i])] = 0
+# map the indices of the left and right joined tables
+# to the indices of the rows in the resulting table
+# if `nothing` is given, the corresponding map is not built
+function update_row_maps!(left_table::AbstractDataTable,
+                          right_table::AbstractDataTable,
+                          right_dict::RowGroupDict,
+                          left_ixs::Union{Void, RowIndexMap},
+                          leftonly_ixs::Union{Void, RowIndexMap},
+                          right_ixs::Union{Void, RowIndexMap},
+                          rightonly_mask::Union{Void, Vector{Bool}})
+    # helper functions
+    @inline update!(ixs::Void, orig_ix::Int, join_ix::Int, count::Int = 1) = nothing
+    @inline function update!(ixs::RowIndexMap, orig_ix::Int, join_ix::Int, count::Int = 1)
+        for i in 1:count
+            push!(ixs.orig, orig_ix)
         end
-    end
-    for i = 1:length(v2)
-        if !_isnull(v2[i])
-            poolref[K(v2[i])] = 0
+        for i in join_ix:(join_ix+count-1)
+            push!(ixs.join, i)
         end
+        ixs
     end
-
-    # fill positions in poolref
-    pool = sort(collect(keys(poolref)))
-    i = 1
-    for p in pool
-        poolref[p] = i
-        i += 1
-    end
-
-    # fill in newrefs
-    zeroval = zero(R)
-    for i = 1:length(v1)
-        if _isnull(v1[i])
-            refs1[i] = zeroval
-        else
-            refs1[i] = poolref[K(v1[i])]
+    @inline update!(ixs::Void, orig_ixs::AbstractArray, join_ix::Int) = nothing
+    @inline function update!(ixs::RowIndexMap, orig_ixs::AbstractArray, join_ix::Int)
+        append!(ixs.orig, orig_ixs)
+        for i in join_ix:(join_ix+length(orig_ixs)-1)
```

Comment:
Should be more efficient as `append!(ixs.join, join_ix:(join_ix+length(orig_ixs)-1))`.